### ISSUE TITLE
Numerically stable derivative for `split_node`

### ISSUE THREAD

**Issue by sidek** [google/TensorNetwork#190]
It would be very nice for TensorNetwork's `net.split_node` and `net.split_node_full_svd` to have a numerically stable derivative out-of-the-box, since this is (I think) the major barrier to TensorNetwork being fully automatically differentiable for practical use. 

Using [this paper's](https://arxiv.org/pdf/1903.09650.pdf) algorithm and [corresponding code](https://github.com/wangleiphy/TRG), I have (very easily) implemented such an SVD for TensorNetwork's PyTorch backend, which works well in calculations, even up to the third derivative. [1] 

I guess the questions I'd like to discuss are:

1. Would it be good to implement numerically stable derivatives for `split_node`, `split_node_full_svd`? Perhaps just for the jax and pytorch backends? 

2. The [algorithm](https://arxiv.org/pdf/1903.09650.pdf) I linked above uses a Lorentzian broadening to make the derivative numerically stable. This introduces some error in the gradient. 

     a) Is there a better algorithm to implement, hopefully one which doesn't introduce error?
    
     b) If the best algorithm is the Lorentzian broadening one, perhaps it's best to add an optional argument `broadening` to `split_node`, maybe with default value 0, so that people can choose an error/stability tradeoff, and are aware that they're making one? That said, having such an argument might make people expect that `split_node` has backend-independent stable derivative... and if that's not the case, it might lead to unexpected errors for TensorFlow users. 

If this seems like a good idea, I'm happy to work on implementing it -- though I can't guarantee that the work will happen in a timely fashion if I do it, and anyone willing to pick this up in a more timely fashion is of course welcome to do so. 

[1]:  When I tried to implement the very same algorithm in the TensorFlow backend, I ran into significant issues (of many kinds) trying to take even the second-order automatic derivative. I haven't tried implementing it in jax, but knowledgeable people have told me that the issues I faced in TensorFlow should not occur there, perhaps another argument for #183 .

**Comment by Thenerdstation** [comment]
I do think this is a good idea!

And yeah tensorflow sucks. I'm not surprised it was hard to get working. 

In general I say go for it! we can discuss how the API should look once we are reasonably certain it's working. (It might also might be a good idea to just include this in the upstream PyTorch and JAX libraries.)

**Comment by amilsted** [comment]
> And yeah tensorflow sucks. I'm not surprised it was hard to get working.

I think I can tell this is not an "official Google product" 😄

**Comment by Bonnevie** [comment]
Just wanted to weigh in with this article on [differentiable linear algebra](https://arxiv.org/abs/1710.08717). They go into a bit of detail/discussion on differentiable matrix operators. Unfortunately they work for Amazon so they only have implementations in MXNet. Remembered it as having an implementation for SVD, but on checking it only has symmetric Eigen, but maybe you can draw something from it anyway.

**Comment by rutujasurve** [comment]
Hi,
I would like to contribute to this issue.

**Comment by Thenerdstation** [comment]
Great! Go for it.

**Comment by rutujasurve** [comment]
Thanks a lot, could you assign me the same?
I am going through Lorentzian broadening and MXNet. Will keep updating with my findings. Any leads wrt previous experiments (how they were run) / results apart from[ torchsvd code](https://github.com/wangleiphy/TRG/blob/master/torchsvd.py) would be very helpful.

**Comment by Thenerdstation** [comment]
Assigned it to you. As for experiments, we don't have any in particular you can run, unless @sidek has something.

**Comment by rutujasurve** [comment]
@sidek @Bonnevie It would be great to hear your inputs on implementing Lorentzian broadening and about a good starting point in the code base where we would introduce the change?
Do we intend on developing wrappers for TensorFlow, jax and PyTorch backends?
